server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java

```diff
@@ -339,57 +270,72 @@ private static Controller selectPost(Proxy proxy, ProxyContext context, String p
                 case "logs" -> controller::getApplicationLogs;
                 default -> null;
             };
-        }
-
-        return null;
-    }
-
-    private static Controller selectDelete(Proxy proxy, ProxyContext context, String path) {
-        Matcher match = match(PATTERN_FILES, path, context);
-        if (match != null) {
+        });
+        // DELETE routes
+        delete(PATTERN_FILES, (proxy, context, pathMatcher) -> {
             DeleteFileController controller = new DeleteFileController(proxy, context);
+            String path = context.getRequest().path();
             return () -> controller.handle(resourcePath(path));
-        }
-
-        match = match(PATTERN_RESOURCE, path, context);
-        if (match != null) {
+        });
+        delete(PATTERN_RESOURCE, (proxy, context, pathMatcher) -> {
             ResourceController controller = new ResourceController(proxy, context, false);
+            String path = context.getRequest().path();
             return () -> controller.handle(resourcePath(path));
-        }
-
-        match = match(INVITATION, path, context);
-        if (match != null) {
-            String invitationId = UrlUtil.decodePath(match.group(1));
+        });
+        delete(INVITATION, (proxy, context, pathMatcher) -> {
+            String invitationId = UrlUtil.decodePath(pathMatcher.group(1));
             InvitationController controller = new InvitationController(proxy, context);
             return () -> controller.deleteInvitation(invitationId);
-        }
-
-        return null;
-    }
-
-    private static Controller selectPut(Proxy proxy, ProxyContext context, String path) {
-        Matcher match = match(PATTERN_FILES, path, context);
-        if (match != null) {
+        });
+        // PUT routes
+        put(PATTERN_FILES, (proxy, context, pathMatcher) -> {
             UploadFileController controller = new UploadFileController(proxy, context);
+            String path = context.getRequest().path();
             return () -> controller.handle(resourcePath(path));
-        }
-
-        match = match(PATTERN_RESOURCE, path, context);
-        if (match != null) {
+        });
+        put(PATTERN_RESOURCE, (proxy, context, pathMatcher) -> {
             ResourceController controller = new ResourceController(proxy, context, false);
+            String path = context.getRequest().path();
             return () -> controller.handle(resourcePath(path));
-        }
+        });
+    }
 
-        return null;
+    public ControllerTemplate select(HttpServerRequest request) {
+        HttpMethod method = request.method();
+        String path = request.path();
+        if (request instanceof HttpServerRequestInternal req) {
```

Comment:
Where do you use it?

Reply:
Seems like you just want to cache the result in tracer and the following code that selects the controller.